### PR TITLE
Binary classifier

### PR DESCRIPTION
**Pull request by tumsifu1**
Binary classifier is ready to be merged in along with the tools and dataset classes.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -162,4 +162,8 @@ cython_debug/
 #.idea/
 
 # Source Video
-data/test.mp4
\ No newline at end of file
+data/test.mp4
+
+data/images/
```
Comment (D-Joseph) [inline .gitignore:167]: ignoring the image and labels, how will we access them? Do we have a 'final' dataset somewhere?

File: models/snoutNet/model.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import torch.nn as nn
+from torch import Tensor
+import torch
+import torch.nn.functional as F
+
+class SnoutNet(nn.Module):
```
Comment (D-Joseph) [inline models/snoutNet/model.py:6]: This class name doesn't make sense in the context of this project lmao

File: models/snoutNet/model.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        # print("Shape After FC1:", X.shape)
+        X = self.fc2(X)
+        X = F.relu(X)
+        # print("Shape After FC2:", X.shape)
+        X = self.fc3(X)
+        # print("Shape After FC3:", X.shape)
+
+        return X
+
+if __name__ == "__main__":
```
Comment (D-Joseph) [inline models/snoutNet/model.py]: Can omit, but no harm done ig

File: models/BallLocalization/FoosballDatasetLocalizer.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    def collate_fn(batch):
+        # Unpack the batch
+        positive_regions = torch.stack([item[0] for item in batch], dim=0)  # Stack all positive_region tensors
+        x_coords = torch.tensor([item[1] for item in batch], dtype=torch.float32)  # Collect all new_x values
+        y_coords = torch.tensor([item[2] for item in batch], dtype=torch.float32)  # Collect all new_y values
+        #ball_exists = torch.tensor([item[3] for item in batch], dtype=torch.float32)  # Collect all ball_exists values
+
+        return positive_regions, x_coords, y_coords
+
+        return regions, x_coords, y_coords
```
Comment (D-Joseph) [inline models/BallLocalization/FoosballDatasetLocalizer.py]:
```suggestion
```

File: models/BallLocalization/FoosballDatasetLocalizer.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    def get_new_coordinates(self, x, y,region_width, region_height):
```
Comment (D-Joseph) [inline models/BallLocalization/FoosballDatasetLocalizer.py:19]: Is this converting the global coordinates to the local coordinates for a specific region? I might figure it out later, but how are we persisting the region of interest data? As in, each segment will have the coordinates (10, 10) for example, but only one segment actually contains the ball at that location

File: models/binaryClassifier/FoosballDataset.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+        """turns data for getitem to tuple of tensors"""
+        #flatten paired regions and labels
+
+        regions = torch.cat([item['regions'] for item in batch], dim=0)
+        labels = torch.cat([item['labels'] for item in batch], dim=0)
+
+        #shuffle regions and labels
+        combined = list(zip(regions, labels))
+        #random.shuffle(combined)
+        regions, labels = zip(*combined)
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:71]: If we are not shuffling (which we don't need to, since the dataloader ideally handles), we can remove these lines with zipping and unpacking
```suggestion

```

File: models/binaryClassifier/FoosballDataset.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            col_index = ball_x // region_width
+            #print(f"Col Index: {col_index} ball_x: {ball_x} region_width: {region_width}")
+
+            # Calculate row_index (this is correct)
+            row_index = ball_y // region_height
+            #print(f"Row Index: {row_index} ball_y: {ball_y} region_height: {region_height}")
+            # Ensure row_index and col_index are within bounds
+            row_index = min(row_index, self.GRID_SIZE - 1)
+            col_index = min(col_index, self.GRID_SIZE - 1)
+
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:100]: Why do we do this? If the indices are out of bounds we should throw it away instead of give it working conditions. I assume this is to prevent errors, but we should be handling that better

File: models/binaryClassifier/FoosballDataset.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            # Calculate region_index
+            region_index = row_index * self.GRID_SIZE + col_index
+
+            #print(f"self.GRID_SIZE: {self.GRID_SIZE}, Region Index: {region_index}")
+
+            if region_index < 0 or region_index >= len(regions):
+                raise IndexError(f"Invalid region_index: {region_index}. Valid range is 0 to {len(regions)-1}.")
+
+            positive_region = regions[region_index]
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:109]: Scratch my last comment, but this should above and should check if the row or col idxs are > self.grid_size

File: models/binaryClassifier/FoosballDataset.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            raise ValueError("Ball does not exist in the image.")
+        
+        random_negative_index = np.random.choice(negative_indices)
+        negative_region = regions[random_negative_index]
+        return negative_region
+    
+    def breakImageIntoRegions(self, image):
+        """Break the image into regions and return a list of regions"""
+        _, height, width = image.shape
+        assert height % self.GRID_SIZE == 0, "Image height is not divisible by grid_size"
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:133]: assert is crazy work I didn't even realize this was a keyword lmao

File: models/binaryClassifier/FoosballDataset.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        assert width % self.GRID_SIZE == 0, "Image width is not divisible by grid_size"
+
+        # Divide the image into regions
+        # Divide the image into regions
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:137]:
```suggestion

```

File: models/binaryClassifier/FoosballDataset.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        region_height = height // self.GRID_SIZE
+        region_width = width // self.GRID_SIZE
+
+        # Divide the image into regions
+        # Divide the image into regions
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:142]:
```suggestion
```

File: models/binaryClassifier/FoosballDataset.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+import random
+import sys
+import os
+from src.tools.unnormalize import unnormalize
+
+class FoosballDataset(Dataset): 
+    def __init__(self, images_dir, json_path, transform=None, train=True):
+
+        self.train = train
+        self.GRID_SIZE = 4 # 4x4 grid
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:21]: will anything break if we make this a parameter instead of harcoding it to 4? May help us with some iterations

File: models/binaryClassifier/FoosballDataset.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+        positive_label = 1 if ball_exists else 0
+        negative_label = 0
+
+        # Return positive and negative samples
+        return {
+            "regions": torch.stack([positive_region, negative_region]),
+            "labels": torch.tensor([positive_label, negative_label]),
+        }
+    
+    #2304 × 1296
```
Comment (D-Joseph) [inline models/binaryClassifier/FoosballDataset.py:190]:
```suggestion
```

File: models/BallLocalization/train.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    optimizer = kwargs["optimizer"]
+    scheduler = kwargs["scheduler"]
+    loss_function = kwargs["loss_function"]
+    train_loader = kwargs["train"]
+    test_loader = kwargs["test"]
+    val_loader = kwargs["val"]
+    device = kwargs["device"]
+    output_dir = kwargs["output"]
+
+    #lists for trianign and val loss
```
Comment (D-Joseph) [inline models/BallLocalization/train.py:29]:
```suggestion
    #lists for training and val loss
```

File: models/BallLocalization/train.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+        with open(file_path, "a") as f:
+            f.write(f"Epoch {epoch+1}/{epochs} | Train Loss: {train_loss:.5f} | Val Loss: {val_loss:.5f} | MSE: {mse_average:.5f} | lr: {optimizer.param_groups[0]['lr']}\n")
+        
+        print(f"Epoch {epoch+1}/{epochs} | Train Loss: {train_loss:.5f} | Val Loss: {val_loss:.5f} | MSE: {mse_average:.5f}")
+
+        torch.save(model.state_dict(), f"{output_dir}/model_epoch_{epoch+1}.pth")
+
+
+        if val_loss == min(losses_val):
+            torch.save(model.state_dict(), os.path.join(output_dir, "best_model.pth"))
```
Comment (D-Joseph) [inline models/BallLocalization/train.py:89]: bars

File: models/BallLocalization/train.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+        plt.xlabel('Epoch')
+        plt.ylabel('Loss')
+        plt.legend()
+        plt.savefig(f"{output_dir}/loss_plot.png")
+        plt.close()
+
+    print("Training complete.")
+    print("starting testing...")
+
+    #testing
```
Comment (D-Joseph) [inline models/BallLocalization/train.py:105]: I didn't realize that we are spllitting the dataset into train test and val. Do we have enough images to have a dedicated validation set? Shouldn't we just use the test set as our val? Also would allow us to remove this additional code

File: models/BallLocalization/train.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    if torch.cuda.is_available():
+        torch.cuda.manual_seed(42)
+        
+    #make sure output folder exists 
+    if not os.path.exists(args.output):
+        os.makedirs(args.output)
+
+    #check if cuda is available
+    device = 'cpu'
+    if torch.cuda.is_available():
```
Comment (D-Joseph) [inline models/BallLocalization/train.py:160]: Add to above, same if condition

File: models/Tests/testDataset.py [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        ball_exists, x, y, regions, region_height, region_width
+    )
+
+    # Assertions
+    assert region_index == 10, f"Expected region_index to be 0, got {region_index}"
+    assert positive_region == regions[10], f"Expected positive_region to be {regions[4]}, got {positive_region}"
+
+    print("Test passed: getRegionWithBall correctly identified the region.")
+
+def __main__():
```
Comment (D-Joseph) [inline models/Tests/testDataset.py]: just Just call this main, not __main__

File: models/BallLocalization/train.py (hunk 16 above)
Comment (D-Joseph) [reply models/BallLocalization/train.py:105]: Regardless, we should test outside of this script to speed up training time

File: models/binaryClassifier/train.py [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+def main():
+    argParser = argparse.ArgumentParser()
+    argParser.add_argument("-epoch", type=int, default=30, help="number of epochs")
+    argParser.add_argument('-images', metavar='images_directory', type=str, help='path to images directory (default: ./data/image_data/images)', default='./data/images')
+    argParser.add_argument('-labels', metavar='labels', type=str, help='path to labels directory', default='./data/labels/labels.json')
+    argParser.add_argument('-batch', metavar='batch_size', type=int, help='batch size, defaults to 64', default=32)
+    argParser.add_argument('-output', metavar='output', type=str, help='output directory', default='./output/binary_classifier')
+    args = argParser.parse_args()
+
+    #random seed for reproducibility
```
Comment (D-Joseph) [inline models/binaryClassifier/train.py:169]: Should remove this moving forward

**Review by D-Joseph — COMMENTED** [review]
Added some comments and some clarification qs

File: models/snoutNet/model.py (hunk 2 above)
Comment (tumsifu1) [reply models/snoutNet/model.py:6]: I added because at first I was going to use it as the base model and was gonna import. Did not end up doing that but kept it in there incase we thought we should use it

File: .gitignore (hunk 1 above)
Comment (tumsifu1) [reply .gitignore:167]: final dataset will be in sharepoint once we can clean it but we can also generate it whenever using our scripts. Didn't want to commit things too large.

File: models/BallLocalization/train.py (hunk 16 above)
Comment (tumsifu1) [reply models/BallLocalization/train.py:105]: Testing doesn't really take any amount of time to do that. Like might add an extra few minutes but I think that's chill overall because we generally train models and leave them it's kinda nice to come back with all the results you need. At least it was for me over the break. 

We do need val and test. We should have done this for 475 in hindsight. Basically our model might become overfit on val dataset since we tune hyper parameters with it. Here is [ a reddit thread ](https://www.reddit.com/r/learnmachinelearning/comments/vogqe5/is_validation_dataset_really_a_must/)that helped me understand

File: models/BallLocalization/train.py (hunk 17 above)
Comment (tumsifu1) [reply models/BallLocalization/train.py:160]: smart

File: models/BallLocalization/FoosballDatasetLocalizer.py (hunk 5 above)
Comment (tumsifu1) [reply models/BallLocalization/FoosballDatasetLocalizer.py:19]: Yes its converting global to local. The region of interest is found elsewhere with a different function. Once we find the index we use that to persist the data.

File: models/binaryClassifier/train.py (hunk 19 above)
Comment (tumsifu1) [reply models/binaryClassifier/train.py:169]: not clear what is meant let's discuss tomorrow

File: models/Tests/testDataset.py (hunk 18 above)
Comment (tumsifu1) [reply models/Tests/testDataset.py]: good catch
